---
title: "HW2MuZepeng"
author: "Zepeng Mu"
date: "10/6/2017"
output: html_document
---

##Lab 2 - Homework

<span style="color:red"> The homework is due by 11:59 PM on the Sunday following your lab. Provide your answers in an Rmd file with the contents of this homework. Call the Rmd file "HW2LastnameFirstname.Rmd" and submit it on Chalk for your T.A. to grade. Remember to comment your code as we did for Lab1. \span

###Part 1 - Data Handling

To prepare for these homework exercises, load the golub data and create factors for ALL versus AML status by repeating the commands from your Lab 2 assignment.

```{r}
#load golub data and generate factor
library(multtest)
data(golub)
gol.fac <- factor(golub.cl, levels=c(0,1), labels = c("ALL","AML"))
```

![](./Images/img4.png)

<span style="color:green"> 1.1 Use the `boxplot()` function in an analogous manner to the stripchart from lab exercise 2.2 to produce a figure similar to the one above. What additional information does a box plot convey? In which situations might a strip chart be preferred, and when might you choose to make a boxplot? \span

<span style="color:blue"> My answer here:  
```{r}
boxplot(golub[1042, ] ~ gol.fac, xlab = "Patient cancer subtype", ylab = "Normalized cyclin D3 expression", main = "My plot")
```

<span style="color:blue">
Boxplots give us more information about the statistic of our data. It shows the first and third quantile of the data, as well as the max/min value (after excluding the outliers) and the median value of the data. So a boxplot may be desired when we want to visualize some of the key statistics of a population. But boxplot cannot show the number of samples in each group, so it might be better to use stripchart when we interested the number of samples. \span

It would be useful to directly visualize the distributions of cyclin D3 expression for ALL and AML patients on the same graphic in order to observe the overlap in their distributions. One way to do this is as follows:
```{r}
hist(golub[1042,gol.fac=="AML"], freq=FALSE, col=4, xlim=c(-1,3), 
     xlab = "Normalized cyclin D3 expression", 
     main="Distribution of normalized cyclin D3 expression in leukemia patients")
hist(golub[1042,gol.fac=="ALL"], freq=FALSE, col=3, add=TRUE)
legend(x=-1,y=0.7, legend=c("ALL","AML"),pch=15, col=c(3,4))
```

As you can see, plotting two histograms in this fashion is difficult—regardless of the order in which you add the distributions, one distribution will crop out the other in certain bins. Let’s explore an alternative method for illustrating these two distributions.  

The `plot()` function that we used in lab can take more than just vectors of values as an input. There are functions in R that transform data in a particular way and the results of these functions can then be passed to the `plot()` function. An example of this is the kernel density function, `density()`. Executing the `density()` command in the following manner will describe a kernel density description of the cyclin D3 gene expression values for ALL patients:  

```{r}
density(golub[1042,gol.fac=="ALL"])
```

<span style="color:green"> 1.2 Use this command in conjunction with `plot()` to create a graphic illustrating the overlap in distributions of cyclin D3 expression between ALL and AML patients like that below. After a plot window has been opened with the first set of data, a second distribution can be layered over the plot with the `lines()` function, which has the same usage as `plot()`. Finally, use the legend command to place a legend on your plot illustrating which color corresponds to which leukemia class. See the histogram example above for the usage of xlim and legend. \span

![](./Images/img5.png)

<span style="color:blue"> My answer here: \span

```{r}
denALL = density(golub[1042,gol.fac=="ALL"])
denAML = density(golub[1042,gol.fac=="AML"])

plot(denALL$x, denALL$y, xlim = c(-1, 3), type = 'l', col = "green", xlab = "Normalized cyclin D3 expression", ylab = "Density", main = "My plot: Distribution of normalized cyclin D3 expression in leukemia patients")
lines(denAML$x, denAML$y, col = "blue")
legend(-1, 1.1, legend = c("ALL", "AML"), col = c("green", "blue"), lty = 1)
```

###Part 2: Finding *oriC*

The next question tries to use the tools you have learned and developed to try to address the question: Where is the origin of replication?

We will try to do this in the bacterial genome, which usually has only one chromosome. The bacterial genome, different from that of animals, plants, fungi and archea, is usually circular. This creates certain subtleties to its analysis, since there isn't a point where it starts or ends, and it isn't easy to handle circular data. Hence, we are going to handle bacterial genomes as DNA sequences, as if the circular genome had been cut at an arbitrary point. Replication is fundamental for the maintenance of life, and hence, finding the point where it starts in the genome might be key to understanding certain life processes. DNA replication is performed by enzymes called DNA polymerases. They do this by first binding to specific regions of DNA, then splitting the two strands of DNA and finally creating new strands, one for each of the original strands. This way, DNA replication is semi-conservative. Our question is then, where does replication start? We will call the region where replication starts as *oriC*. There are ways of finding *oriC* experimentally, but these methods are much slower. If we can find, using computational methods, a way of suggesting some regions where *oriC* could be, even though not giving the exact region, we might help a lot in speeding up these experimental approaches.  

One approach to finding *oriC* would be to try to find some words in the bacterial genome that indicate that region as where replication starts. This could be done statistically, by finding unusually frequent words (in the sense that it is more frequent in the DNA genome than in a randomly generated sequence of the letters "A", "G", "C" and "T" of the same length) clumping together within short regions of the genome. We are going to instead use an approach based on the content of guanine and the content of cytosine in the genome, by using a sliding window approach similar to next weeks lab.  

But why should the frequencies of G and C help settle this question? This was hinted at in the lab. It turns out that replication generates mutational anomalies due to the asymmetry of replication. This is because DNA polymerase is *unidirectional*, it can only transverse DNA in the reverse direction (from 3' to 5'). This causes no problem for the reverse half-strand, but the forward half-strand, which is unraveled for replication from 5' to 3', can't be replicated from the start; it must wait until there is enough space for the DNA polymerase (these enzymes are huge biological machines, and need a fair amount of space before binding) to bind a few nucleotides down the road and then be able to replicate the short segment that was unraveled thus far in the direction 3'-> 5'.
These generates several phenomena in DNA, such as the famous Okazaki fragments. What is relevant to us though, is that, because of this, the forward half-strand spends much more time single stranded than the reverse half-strand. And single-stranded DNA has much higher mutation rates than double-stranded DNA! Specifically, C tends to mutate into T by a process called deamination. The rate of deamination in single-stranded DNA is 100 times higher than in double-stranded DNA. This creates mismatched pairs T-G which are then corrected in the next replication into T-A pairs. It should be the case then, that in the forward half-strand the G content should be low in relation to the C content, and the opposite behavior in the reverse half-strand (because the forward half-strand of the complementary strand pairs with the reverse half-strand of the original strand). In particular, `G_content(genome)-C_content(genome)`, for some genome, should be smallest close to the *oriC*, since *oriC* is the point that separates the beginning of the reverse half-strand and the end of the forward half-strand. The following code loads a package which we will use in the next questions.

```{r}
#install.packages("seqinr", repos='http://cran.us.r-project.org')
library("seqinr")
```

<span style="color:green"> 2.1 With all of this said, let's analyze the most studied bacterial genome: *E. Coli*. \span

<span style="color:green"> (a) Put the *E. coli*, *Thermotoga petrophila* and *Sulfolobus solfataricus* in your working directory. The following code loads all of these files into R. You will learn how to do all of this by yourself (including finding the genomes) in the next lab. \span  

```{r}
e_coli <- read.fasta("E. coli.fasta")[[1]]
t_petrophila <- read.fasta("Thermotoga petrophila.fasta")[[1]]
s_solfataricus <- read.fasta("Sulfolobus solfataricus.fasta")[[1]]
```
<span style="color:green"> (b) What we will do is try to find the point where the difference between the G content and the C content achieves a minimum. For that, write a function called `skew()` that receives as input a DNA sequence (which in our case will be in fact a whole bacterial genome) and a positive integer i and outputs the difference between the G count and the C count from the first nucleotide in the DNA sequence up to nucleotide number i (for i = 0, define the skew to be 0).  \span

<span style="color:blue"> My code here: \span

```{r}
skew = function(inSeq, i) {
  
  sk = 0

  if(i == 0) {
    return(0)
  } else {
    for(nt in inSeq[1:i]) {
      if(nt == "G") {
        sk = sk + 1
      } else if(nt == "C") {
        sk = sk - 1
      }
    }
    return(sk)
    
  }
}
```

<span style="color:green"> (c) Now, write a function called `skewDiagram()` that, given a DNA sequence, outputs the skew diagram of the DNA sequence. A skew diagram is a plot that has on the x axis positive integers (with x ranging from 0 to the length of the DNA sequence), and on the y axis the skew of the DNA sequence up to the index x. Hint: This is the first time in this course you are going to face genomical analysis. One of its challenges is that genomes are huge. Even a bacterial genome might be non-trivial to a personal computer. Therefore, your implementation of the algorithm might affect a lot the time you spend in this lab. Think carefully on how to implement your function. First of all, do not recount G and C every time you need to compute some skew. Instead, find a way of just updating previous skews. (**NOTE: to make this execute in << tens of minutes, plot `skew[seq(1, length(skew), 10000)]` vs. `seq(1, length(skew), 10000)`) Be sure to ask to your TA for help if needed.\span 

<span style="color:blue"> My code here: \span

```{r}
skewDiagram = function(inSeq) {
  
  skewVec = vector(length = length(inSeq))
  
  sk = 0
  gc = 0

  for(i in 1:length(inSeq)) {
    if(inSeq[i] == "g" || inSeq[i] == "G") {
      sk = sk + 1
      skewVec[i] = sk
    } else if(inSeq[i] == "c" || inSeq[i] == "C") {
      sk = sk - 1
      skewVec[i] = sk
    } else {
      skewVec[i] = sk
    }
  }
  
  x = seq(1, length(skewVec), 10000)
  y = skewVec[seq(1, length(skewVec), 10000)]
  plot(x, y, type = "l", xlab = "genome position", ylab = "skew value")
}
```

<span style="color:green"> (d) Write a function `min_skew` that, given a DNA sequence, finds the i such that `skew(DNA, i)` is the minimum among possible i. \span

<span style="color:blue"> My code here: \span

```{r}
min_skew = function(inSeq) {
  
  skewVec = vector(length = length(inSeq))
  
  sk = 0

  for(i in 1:length(inSeq)) {
    if(inSeq[i] == "g" || inSeq[i] == "G") {
      sk = sk + 1
      skewVec[i] = sk
    } else if(inSeq[i] == "c" || inSeq[i] == "C") {
      sk = sk - 1
      skewVec[i] = sk
    } else {
      skewVec[i] = sk
    }
  }
  
  return(order(skewVec)[1])
  
}
```

<span style="color:green"> (e) Now, use the functions you just wrote to analyze the genome of *E. coli*. Given the skew diagram of *E. coli*, report **and justify** where you expect *oriC* to be located in the *E. coli* genome (notice that the minimum may differ from *oriC* due to random fluctuations in the G or C frequencies, hence, you should really indicate a short region rather than a point). (Hint: your suggested region should be a span 500 to 1000 nucleotides long) \span

<span style="color:blue"> My code here: \span

```{r}
skewDiagram(e_coli)

#min_skew(e_coli) #3925597
```

<span style="color:blue"> From the diagram above, we see that the skew value of *E. coli* genome reaches an overall minimum close to 400,0000 on the genome. Running `min_skew()` function returns 3925597. Therefore, I would like to suggest that the *oriC* of *E. coli* genome lies within the region of 3925000 - 3926000. \span

<span style="color:green"> 2.2 Now, load the genome of *Thermotoga petrophila* and plot its skew diagram. How is this different from that of *E. coli*? Where would you expect the *oriC* of *Thermotoga petrophila* to be? \span

<span style="color:blue"> My code here: \span

```{r}
skewDiagram(t_petrophila)

#min_skew(t_petrophila) #787199
```

<span style="color:blue"> From the diagram above, we see that the skew value of *Thermotoga petrophila* genome reaches an overall minimum close to 400,0000 on the genome. Running `min_skew()` function returns 787199. Therefore, I would like to suggest that the *oriC* of *Thermotoga petrophila* genome lies within the region of 786500 - 787500. \span

<span style="color:green"> 2.3 Load the genome of the archea *Sulfolobus solfataricus* and plot its skew diagram. We've seen that the minimum skew provides a guess for the *oriC* of bacterial genome. Notice though, that we were focusing on global minima. You can see on the skew diagram of *Sulfolobus solfataricus* that there are three clear valleys were you can identify local minima. What might this indicate? Try to find published experimental evidence that proves your hypothesis. \span

<span style="color:blue"> My code here: \span

```{r}
skewDiagram(s_solfataricus)

#min_skew(s_solfataricus) #749375
```

<span style="color:blue"> From the diagram above, we see that the skew value of *Sulfolobus solfataricus* genome has at least two local minimum. Running `min_skew()` function returns 749375. And there is also a second minimum at the position of around 600000. This might suggest that there is more than one *oriC* in *Sulfolobus solfataricus* genome. In fact, this has been reported before. In a paper publishe in early 2014, it was proved that there were two replication origins in the genome of *Sulfolobus solfataricus* (Identification of Two Origins of Replication in the Single Chromosome of the Archaeon *Sulfolobus solfataricus*, Cell, Vol. 116, 25–38, January 9, 2004). A later work then used cumulative-GC skew to predict the position of *oriC*s in the genome, and researchers identified two positions the same as what I found here (Identification and autonomous replication capability of a chromosomal replication origin from the archaeon *Sulfolobus solfataricus*, Extremophiles (2004) 8:385–391). However, a third *oriC* was not reported by literature. Therefore, I would like to suggest that the *oriC* of *Sulfolobus solfataricus* genome lies within the region around 0.6 Mbp and 0.75 Mbp. \span
